src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,462 @@
+      pred_lpdf_united_clust += united_clust_unique_values->prior_pred_lpdf(
+          data.row(random_idxs[clust_idx]));
+    } else {
+      pred_lpdf_united_clust +=
+          united_clust_unique_values->conditional_pred_lpdf(
+              data.row(random_idxs[clust_idx]));
+      log_ratio_prior_prob -= mixing->get_mass_existing_cluster(
+          allocations.size(), true, true, united_clust_unique_values);
+    }
+    united_clust_unique_values->add_datum(random_idxs[clust_idx],
+                                          data.row(random_idxs[clust_idx]),
+                                          update_hierarchy_params());
+
+    std::set<int> set_to_cycle;
+    std::set<int>::const_iterator it;
+    std::set<int>::const_iterator end;
+    if (split) {
+      set_to_cycle = restricted_gs_unique_values[clust_idx]->get_data_idx();
+      set_to_cycle.erase(random_idxs[clust_idx]);
+      it = set_to_cycle.cbegin();
+      end = set_to_cycle.cend();
+    } else {
+      set_to_cycle =
+          unique_values[allocations[random_idxs[clust_idx]]]->get_data_idx();
+      set_to_cycle.erase(random_idxs[clust_idx]);
+      it = set_to_cycle.cbegin();
+      end = set_to_cycle.cend();
+    }
+
+    for (; it != end; ++it) {
+      unsigned int curr_idx = (*it);
+      pred_lpdf_divided_clust(clust_idx) +=
+          divided_clust_unique_values[clust_idx]->conditional_pred_lpdf(
+              data.row(curr_idx));
+      log_ratio_prior_prob += mixing->get_mass_existing_cluster(
+          allocations.size(), true, true,
+          divided_clust_unique_values[clust_idx]);
+      divided_clust_unique_values[clust_idx]->add_datum(
+          curr_idx, data.row(curr_idx), update_hierarchy_params());
+
+      pred_lpdf_united_clust +=
+          united_clust_unique_values->conditional_pred_lpdf(
+              data.row(curr_idx));
+      log_ratio_prior_prob -= mixing->get_mass_existing_cluster(
+          allocations.size(), true, true, united_clust_unique_values);
+      united_clust_unique_values->add_datum(curr_idx, data.row(curr_idx),
+                                            update_hierarchy_params());
+    }
+  }
+
+  if (!split) {
+    log_ratio_prior_prob = -log_ratio_prior_prob;
+  }
+
+  log_ratio_likelihoods =
+      pred_lpdf_divided_clust.sum() - pred_lpdf_united_clust;
+  if (!split) {
+    log_ratio_likelihoods = -log_ratio_likelihoods;
+  }
+}
+
+void SplitAndMergeAlgorithm::split_or_merge(
+    const unsigned int first_random_idx,
+    const unsigned int second_random_idx) {
+  if (allocations[first_random_idx] == allocations[second_random_idx]) {
+    split(first_random_idx, second_random_idx);
+  } else {
+    merge(first_random_idx, second_random_idx);
+  }
+}
+
+void SplitAndMergeAlgorithm::split(const unsigned int first_random_idx,
+                                   const unsigned int second_random_idx) {
+  double log_ratio_transition_prob = -restricted_gibbs_sampling(true);
```

Comment:
I think that this call can be avoided since you call `split` after you have already called  `restricted_gibbs_sampling`